File: ui/packages/core/src/webauthn.ts
/**
 * WebAuthn encoding helpers.
 *
 * The go-webauthn backend serialises binary fields (challenge, user.id,
 * credential IDs, …) as **base64url strings** in JSON.  The browser
 * WebAuthn API requires those same fields as ArrayBuffer / Uint8Array.
 *
 * These helpers bridge the gap in both directions:
 *   begin response  → browser API  (base64url string → ArrayBuffer)
 *   browser API     → finish POST  (ArrayBuffer → base64url string)
 */

// ── low-level codecs ──────────────────────────────────────────────

/** Decode a base64url-encoded string into a Uint8Array. */
export function base64urlToBuffer(base64url: string): Uint8Array {
  const base64 = base64url.replace(/-/g, "+").replace(/_/g, "/");
  const padded = base64 + "=".repeat((4 - (base64.length % 4)) % 4);
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/** Encode an ArrayBuffer (or Uint8Array) to a base64url string (no padding). */
export function bufferToBase64url(
  buffer: ArrayBuffer | Uint8Array,
): string {
  const bytes =
    buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

// ── begin → browser helpers ────────────────────────────────────────

/**
 * Prepare the options object returned by `/passkeys/register/begin`
 * for `navigator.credentials.create({ publicKey })`.
 *
 * Converts base64url strings → Uint8Array for:
 *   challenge, user.id, excludeCredentials[].id
 */
export function prepareCreationOptions(
  options: Record<string, unknown>,
): PublicKeyCredentialCreationOptions {
  const publicKey = (options.publicKey ?? options) as Record<string, unknown>;

  const out: Record<string, unknown> = { ...publicKey };

  // challenge
  if (typeof out.challenge === "string") {
    out.challenge = base64urlToBuffer(out.challenge);
  }

  // user.id
  if (out.user && typeof out.user === "object") {
    const user = { ...(out.user as Record<string, unknown>) };
    if (typeof user.id === "string") {
      user.id = base64urlToBuffer(user.id);
    }
    out.user = user;
  }

  // excludeCredentials[].id
  if (Array.isArray(out.excludeCredentials)) {
    out.excludeCredentials = (
      out.excludeCredentials as Record<string, unknown>[]
    ).map((c) => {
      if (typeof c.id === "string") {
        return { ...c, id: base64urlToBuffer(c.id) };
      }
      return c;
    });
  }

  return out as unknown as PublicKeyCredentialCreationOptions;
}

/**
 * Prepare the options object returned by `/passkeys/login/begin`
 * for `navigator.credentials.get({ publicKey })`.
 *
 * Converts base64url strings → Uint8Array for:
 *   challenge, allowCredentials[].id
 */
export function prepareRequestOptions(
  options: Record<string, unknown>,
): PublicKeyCredentialRequestOptions {
  const publicKey = (options.publicKey ?? options) as Record<string, unknown>;

  const out: Record<string, unknown> = { ...publicKey };

  // challenge
  if (typeof out.challenge === "string") {
    out.challenge = base64urlToBuffer(out.challenge);
  }

  // allowCredentials[].id
  if (Array.isArray(out.allowCredentials)) {
    out.allowCredentials = (
      out.allowCredentials as Record<string, unknown>[]
    ).map((c) => {
      if (typeof c.id === "string") {
        return { ...c, id: base64urlToBuffer(c.id) };
      }
      return c;
    });
  }

  return out as unknown as PublicKeyCredentialRequestOptions;
}

// ── browser → finish helpers ───────────────────────────────────────

/**
 * Serialize a PublicKeyCredential returned by navigator.credentials
 * into a plain object suitable for POSTing to the `/finish` endpoint.
 *
 * Converts ArrayBuffer fields → base64url strings so they survive
 * JSON.stringify and match what go-webauthn expects.
 */
export function serializeCredential(
  credential: Credential,
): Record<string, unknown> {
  const cred = credential as PublicKeyCredential;
  const response = cred.response;

  const result: Record<string, unknown> = {
    id: cred.id,
    rawId: bufferToBase64url(cred.rawId),
    type: cred.type,
    response: {} as Record<string, unknown>,
  };

  const serializedResponse: Record<string, unknown> = {
    clientDataJSON: bufferToBase64url(response.clientDataJSON),
  };

  // Attestation response (register finish)
  if ("attestationObject" in response) {
    const attestation = response as AuthenticatorAttestationResponse;
    serializedResponse.attestationObject = bufferToBase64url(
      attestation.attestationObject,
    );
  }

  // Assertion response (login finish)
  if ("authenticatorData" in response) {
    const assertion = response as AuthenticatorAssertionResponse;
    serializedResponse.authenticatorData = bufferToBase64url(
      assertion.authenticatorData,
    );
    serializedResponse.signature = bufferToBase64url(assertion.signature);
    if (assertion.userHandle) {
      serializedResponse.userHandle = bufferToBase64url(assertion.userHandle);
    }
  }

  result.response = serializedResponse;

  // Include authenticatorAttachment if present
  if (cred.authenticatorAttachment) {
    result.authenticatorAttachment = cred.authenticatorAttachment;
  }

  return result;
}

File: flutter/packages/authsome_core/lib/src/webauthn.dart
/// WebAuthn encoding helpers.
///
/// The go-webauthn backend serialises binary fields (challenge, user.id,
/// credential IDs, …) as **base64url strings** in JSON. The browser
/// WebAuthn API (and `passkeys` corbado package on native) require those
/// same fields as raw bytes (`Uint8List` / `ArrayBuffer`).
///
/// These helpers bridge the gap in both directions:
///   begin response  → platform API  (base64url string → Uint8List)
///   platform API    → finish POST   (Uint8List → base64url string)
///
/// Ports `ui/packages/core/src/webauthn.ts` 1:1.
library;

import 'dart:convert';
import 'dart:typed_data';

// ── low-level codecs ──────────────────────────────────────────────

/// Decode a base64url-encoded string into a [Uint8List].
Uint8List base64urlToBytes(String base64url) {
  // Restore the standard alphabet and re-pad before calling `base64.decode`.
  var s = base64url.replaceAll('-', '+').replaceAll('_', '/');
  final pad = (4 - (s.length % 4)) % 4;
  s = s + ('=' * pad);
  return base64.decode(s);
}

/// Encode bytes to a base64url string (no padding).
String bytesToBase64url(Uint8List bytes) {
  return base64
      .encode(bytes)
      .replaceAll('+', '-')
      .replaceAll('/', '_')
      .replaceAll('=', '');
}

// ── begin → platform helpers ───────────────────────────────────────

/// Prepare the options object returned by `/v1/passkeys/login/begin`
/// for `navigator.credentials.get({ publicKey })` (Web) or its native
/// equivalent.
///
/// Converts base64url strings → [Uint8List] for:
///   challenge, allowCredentials[].id
///
/// The server may return options nested under `publicKey` or flat —
/// both shapes are accepted to match the React behaviour.
Map<String, dynamic> prepareRequestOptions(Map<String, dynamic> options) {
  final publicKey = (options['publicKey'] is Map)
      ? Map<String, dynamic>.from(options['publicKey'] as Map)
      : Map<String, dynamic>.from(options);

  final out = <String, dynamic>{...publicKey};

  final challenge = out['challenge'];
  if (challenge is String) {
    out['challenge'] = base64urlToBytes(challenge);
  }

  final allow = out['allowCredentials'];
  if (allow is List) {
    out['allowCredentials'] = allow.map((entry) {
      if (entry is Map) {
        final m = Map<String, dynamic>.from(entry);
        final id = m['id'];
        if (id is String) m['id'] = base64urlToBytes(id);
        return m;
      }
      return entry;
    }).toList();
  }

  return out;
}

/// Prepare the options object returned by `/v1/passkeys/register/begin`
/// for `navigator.credentials.create({ publicKey })`.
///
/// Converts base64url strings → [Uint8List] for:
///   challenge, user.id, excludeCredentials[].id
Map<String, dynamic> prepareCreationOptions(Map<String, dynamic> options) {
  final publicKey = (options['publicKey'] is Map)
      ? Map<String, dynamic>.from(options['publicKey'] as Map)
      : Map<String, dynamic>.from(options);

  final out = <String, dynamic>{...publicKey};

  final challenge = out['challenge'];
  if (challenge is String) {
    out['challenge'] = base64urlToBytes(challenge);
  }

  final user = out['user'];
  if (user is Map) {
    final u = Map<String, dynamic>.from(user);
    final id = u['id'];
    if (id is String) u['id'] = base64urlToBytes(id);
    out['user'] = u;
  }

  final exclude = out['excludeCredentials'];
  if (exclude is List) {
    out['excludeCredentials'] = exclude.map((entry) {
      if (entry is Map) {
        final m = Map<String, dynamic>.from(entry);
        final id = m['id'];
        if (id is String) m['id'] = base64urlToBytes(id);
        return m;
      }
      return entry;
    }).toList();
  }

  return out;
}

// ── platform → finish helpers ──────────────────────────────────────

/// Build the credential JSON payload expected by the `/finish` endpoints.
///
/// Mirrors `serializeCredential` from `webauthn.ts`: takes raw byte fields
/// returned by the platform's WebAuthn API and re-encodes them as
/// base64url strings so they survive `jsonEncode` and match what
/// go-webauthn expects on the wire.
///
/// Inputs are deliberately untyped (`dynamic`) because we accept either
/// [Uint8List] or already-encoded [String] values — handy when bridging
/// from packages that pre-encode (e.g. corbado `passkeys`).
Map<String, dynamic> serializeAssertion({
  required String id,
  required dynamic rawId,
  required String type,
  required dynamic clientDataJson,
  required dynamic authenticatorData,
  required dynamic signature,
  dynamic userHandle,
  String? authenticatorAttachment,
}) {
  String enc(dynamic v) {
    if (v is String) return v;
    if (v is Uint8List) return bytesToBase64url(v);
    if (v is List<int>) return bytesToBase64url(Uint8List.fromList(v));
    throw ArgumentError(
      'Expected Uint8List, List<int>, or base64url String, got '
      '${v.runtimeType}',
    );
  }

  final response = <String, dynamic>{
    'clientDataJSON': enc(clientDataJson),
    'authenticatorData': enc(authenticatorData),
    'signature': enc(signature),
  };
  if (userHandle != null) {
    response['userHandle'] = enc(userHandle);
  }

  final out = <String, dynamic>{
    'id': id,
    'rawId': enc(rawId),
    'type': type,
    'response': response,
  };
  if (authenticatorAttachment != null) {
    out['authenticatorAttachment'] = authenticatorAttachment;
  }
  return out;
}
